File: Nursery/nodejs/odbc-dsnless.js
#!/usr/bin/env nodejs
// tested on Debian using external NodeJS repo with NodeJS 10.0

const odbc = require('odbc');

// Driver : on Linux check with odbcinst -q -d
// const connectionString = "DRIVER={FreeTDS};SERVER=host;UID=user;PWD=password;DATABASE=dbname";
// const connectionString = 'Driver=PervasiveSQL;ServerName=localhost;dbq=@demodata;';
var connectionString = `${process.env.CONNECTION_STRING}`; // must be set in shell script calling this code

const queryString = 'SELECT COUNT(*) FROM PERSON';

console.log('About to make a DSN-less ODBC connection using the following connection-string: %s', connectionString); 

// const connection = odbc.connect(connectionString, (error, connection) => {
const connection = odbc.connect(`${process.env.CONNECTION_STRING}`, (error, connection) => {
    if (error) { console.error(error) }
    connection.query(queryString, (error, result) => {
	if (error) { console.error(error) }
	console.log('Result of query: %s :', queryString);
	var personcount = result[0];
	console.log(personcount);
    });
connection.close();
});

File: ruby/odbc/odbc-dsnless.rb
#!/usr/bin/env ruby

require "odbc"

# http://www.ch-werner.de/rubyodbc/
# https://github.com/localytics/odbc_adapter
# http://www.iodbc.org/dataspace/doc/iodbc/wiki/iodbcWiki/IODBCRubyHOWTO
# https://www.easysoft.com/developer/languages/ruby/rails.html
# https://engineering.ezcater.com/the-long-long-journey-of-connecting-to-snowflake-with-ruby

#$driver='Pervasive ODBC Client Interface'
$driver='Pervasive ODBC Interface'
$server='localhost'
$dbuser=''
$dbpass=''
$dbname='demodata'

$query = 'SELECT COUNT(*) FROM Person'

# https://www.connectionstrings.com/pervasive/
# windows: see ODBC Manager
# linux (unixodbc): odbcinst -q -d and odbcinst -q -s
# mac: see linux?

puts "Ruby ODBC access using native ODBC : DSN-less" + "\n"

# DNS-less coonnection
connectionstring = "DRIVER={" + $driver + "};SERVER=" + $server + ";dbq=@" + $dbname + ";"

puts "Connectstring: #{connectionstring}" + "\n"

# http://www.ch-werner.de/rubyodbc/odbc.html
con = ODBC::Database.new.drvconnect(connectionstring) or die "Unable to connect: " + ODBC::error + "\n"

puts "Connected to: " + con.connected?.to_s + "\n"

# count elements in recordset
stmt = con.run($query) or die "Unable to execute query: " + ODBC::error + "\n"
rs = stmt.fetch_all
$cnt = rs[0]

puts "Row count on table Person: #{$cnt} " + "\n"

con.disconnect